codelabs/custom_generator/custom_generator.md

```diff
@@ -0,0 +1,379 @@
+};
+```
+
+### Number
+
+The `math_number` block has a number field. Like the `text` block, you can use `getFieldValue`. Unlike the text block, you don't need to wrap it in additional quotation marks.
+
+```js
+myGenerator['math_number'] = function(block) {
+  const code = Number(block.getFieldValue('NUM'));
+  return [code, myGenerator.PRECEDENCE];
+};
+```
+
+### Boolean
+
+The `logic_boolean` block has a dropdown field named `BOOL`. Calling `getFieldValue` on a dropdown field returns the value of the selected option, which may not be the same as the display text. In this case the dropdown has two possible values: `TRUE` and `FALSE`.
+
+```js
+myGenerator['logic_boolean'] = function(block) {
+  const code = (block.getFieldValue('BOOL') == 'TRUE') ? 'true' : 'false';
+  return [code, myGenerator.PRECEDENCE];
+```
+
+### Summary
+
+- Value blocks return an array containing the value as a string and the precedence.
+- `getFieldValue` find the field with the specified name and returns its value.
+- The type of the return value from `getFieldValue` depends on the type of the field.
+  - Each field type must document what its value represents.
+
+
+## Member block generator
+
+In this step you will build the generator for the `member` block. You will use `getFieldValue`, and add `valueToCode` to your tool kit.
+
+The member block has a text input field and a value input. It generates code that looks like `"a": b,`.
+
+
+### Field value
+`a` is the value of the text input, which we get with `getFieldValue`:
+
+```js
+ const text_member_name = block.getFieldValue('MEMBER_NAME');
+```
+
+### Input value
+`b` is whatever is attached to the value input. A variety of blocks could be attached there: string, number, boolean, null, or even an array. Use `valueToCode` to get the correct value:
+
+```js
+const value_name = codelabGenerator.valueToCode(block, 'MEMBER_VALUE',
+    codelabGenerator.PRECEDENCE);
+```
+
+`valueToCode` does three things:
+- Finds the blocks connected to the named value input (the second argument)
+- Generates the code for that block
+- Returns the code as a string
+
+If no block is attached, `valueToCode` returns `null`.
+
+The third argument is related to operator precedence, as discussed in a previous section.
+
+### Build the code string
+Next, assemble the arguments `a` and `b` into the correct code, of the form `"a": b,`.
+
+To generate clean code, you should also add a newline at the end of the statement.
+
+```js
+const code = '"' + text_member_name + '" : ' + value_name + ',\n'
+```
+
+### Put it all together
+
+All together, the block generator for the member block looks like this:
+
+```js
+myGenerator['member'] = function(block) {
+  const text_member_name = block.getFieldValue('MEMBER_NAME');
+  const value_name = codelabGenerator.valueToCode(block, 'MEMBER_VALUE',
+      codelabGenerator.PRECEDENCE);
+  const code = '"' + text_member_name + '" : ' + value_name + ',\n';
+  return code;
+};
+```
+
```

Comment:
Will you add one for myGenerator['object'] ?

Reply:
Done.